### PR TITLE
Refinement criterion for free surface based on slope

### PR DESCRIPTION
**Pull request by ian-r-rose**
Adjust header

### CODE REVIEW

**Comment by tjhei** [comment]
- Please fix astyle.
- I wonder if it makes sense to scale the indicator with h so that larger faces with same slope are refined first.
- Add a changelog entry.

Great otherwise.

File: source/mesh_refinement/slope.cc [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    template <int dim>
+    void
+    Slope<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Model settings");
+      {
+        const std::vector<types::boundary_id> x_free_surface_boundary_indicators
+          = this->get_geometry_model().translate_symbolic_boundary_names_to_ids(Utilities::split_string_list
+                                                                    (prm.get ("Free surface boundary indicators")));
```
Comment (bangerth) [inline source/mesh_refinement/slope.cc]: I'm not too fond of this -- it duplicated information in different places. I'd much rather you not have to store this piece of information in your model here, and instead introduce a function into `SimulatorAccess` that returns the set of free surface boundary indicators. This way you could also get rid of the member variable of the current class.

**Comment by ian-r-rose** [comment]
I am not sure whether in practice it matters having h dependence, as the steeper slope cells will hit the maximum refinement level pretty quickly anyways.  Do you have an idea for what to do there?

**Comment by tjhei** [comment]
Ian, please merge https://github.com/tjhei/aspect/commit/c01561ec1d955a5078d0d03df591f2524d626637

**Comment by ian-r-rose** [comment]
Still failing -- can we get another update to the test results?

File: source/mesh_refinement/slope.cc [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                     this->get_free_surface_boundary_indicators().end() )
+                  {
+                    fe_face_values.reinit(cell, face_no);
+
+                    const Tensor<1,dim> normal( fe_face_values.normal_vector(0) ); //Only one q point
+                    const Point<dim> midpoint = fe_face_values.quadrature_point(0);
+                    const Tensor<1,dim> gravity = this->get_gravity_model().gravity_vector(midpoint);
+
+                    indicators(i) = std::min ( std::acos ( normal * gravity / gravity.norm() ),
+                                               std::acos (-normal * gravity / gravity.norm() ) )  //Don't care whether gravity is in the opposite direction
```
Comment (bangerth) [inline source/mesh_refinement/slope.cc]: Isn't this simply `acos(abs(normal * gravity / gravity.norm())`?

File: source/mesh_refinement/slope.cc [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                                    * std::pow( cell->diameter(), dim-1);
```
Comment (bangerth) [inline source/mesh_refinement/slope.cc]: why this particular power of h?

File: source/mesh_refinement/slope.cc (hunk 2 above)
Comment (ian-r-rose) [inline source/mesh_refinement/slope.cc]: You are right, don't know why I didn't think of that.  Will change.

File: source/mesh_refinement/slope.cc (hunk 3 above)
Comment (ian-r-rose) [inline source/mesh_refinement/slope.cc]: The idea is that this scales with the surface area of the surface cell.  It seems to work okay based on simple tests.

**Comment by tjhei** [comment]
Ian, can you take the last commit from https://github.com/tjhei/aspect/commits/pull-359 please?

**Pull request merged**